This patch replaces the hardcoded "root" superuser name from klaptopdaemon by calls to a
new method named laptop_portable::get_superuser_name(). "root" is still used as fallback
when no user with uid=0 is found in the password entries.

Signed-off-by: Lucas C. Villa Real <lucasvr@gobolinux.org>

diff -Nur kdeutils-3.5.8/klaptopdaemon-orig/acpi.cpp kdeutils-3.5.8/klaptopdaemon/acpi.cpp
--- kdeutils-3.5.8/klaptopdaemon-orig/acpi.cpp	2006-10-01 10:28:13.000000000 -0700
+++ kdeutils-3.5.8/klaptopdaemon/acpi.cpp	2008-01-27 20:40:55.000000000 -0800
@@ -182,11 +182,12 @@
 				i18n("KLaptopDaemon"), KStdGuiItem::cont(),
 				"");
 		if (rc == KMessageBox::Continue) {
+			QString superuser = laptop_portable::get_superuser_name();
 			KProcess proc;
 			proc << kdesu;
 			proc << "-u";
-			proc << "root";
-			proc <<  "chown root "+helper+"; chmod +s "+helper;
+			proc << superuser;
+			proc <<  "chown "+superuser+" "+helper+"; chmod +s "+helper;
 			proc.start(KProcess::Block);	// run it sync so has_acpi below sees the results
 		}
 	} else {
diff -Nur kdeutils-3.5.8/klaptopdaemon-orig/apm.cpp kdeutils-3.5.8/klaptopdaemon/apm.cpp
--- kdeutils-3.5.8/klaptopdaemon-orig/apm.cpp	2006-10-01 10:28:13.000000000 -0700
+++ kdeutils-3.5.8/klaptopdaemon/apm.cpp	2008-01-27 20:41:12.000000000 -0800
@@ -162,11 +162,12 @@
 				"KLaptopDaemon", KStdGuiItem::cont(),
 				"");
 		if (rc == KMessageBox::Continue) {
+			QString superuser = laptop_portable::get_superuser_name();
 			KProcess proc;
 			proc << kdesu;
 			proc << "-u";
-			proc << "root";
-			proc <<  QString("chown root ")+apm_name+"; chmod +s "+apm_name;
+			proc << superuser;
+			proc <<  "chown "+superuser+" "+apm_name+"; chmod +s "+apm_name;
 			proc.start(KProcess::Block);	// run it sync so has_apm below sees the results
 		}
 	} else {
@@ -204,11 +205,12 @@
 				i18n("KLaptopDaemon"), KStdGuiItem::cont(),
 				"");
 		if (rc == KMessageBox::Continue) {
+			QString superuser = laptop_portable::get_superuser_name().ascii();
 			KProcess proc;
 			proc << kdesu;
 			proc << "-u";
-			proc << "root";
-			proc <<  "chown root "+helper+"; chmod +s "+helper;
+			proc << superuser;
+			proc <<  "chown "+superuser+" "+helper+"; chmod +s "+helper;
 			proc.start(KProcess::Block);	// run it sync so has_acpi below sees the results
 		}
 	} else {
diff -Nur kdeutils-3.5.8/klaptopdaemon-orig/daemondock.cpp kdeutils-3.5.8/klaptopdaemon/daemondock.cpp
--- kdeutils-3.5.8/klaptopdaemon-orig/daemondock.cpp	2007-10-08 02:49:45.000000000 -0700
+++ kdeutils-3.5.8/klaptopdaemon/daemondock.cpp	2008-01-27 20:42:25.000000000 -0800
@@ -244,11 +244,12 @@
 				i18n("KLaptopDaemon"), KStdGuiItem::cont(),
 				"switchToPCMCIAPrompt");
 		if (rc == KMessageBox::Continue) {
+			QString superuser = laptop_portable::get_superuser_name();
 			KProcess *_rootProcess;
 			_rootProcess = new KProcess;
 			*_rootProcess << kdesu;
 			*_rootProcess << "-u";
-			*_rootProcess << "root";
+			*_rootProcess << superuser;
 			//*_rootProcess << "--nonewdcop";
 			*_rootProcess << KStandardDirs::findExe("klaptopdaemon");
 			connect(_rootProcess, SIGNAL(processExited(KProcess*)),
diff -Nur kdeutils-3.5.8/klaptopdaemon-orig/portable.cpp kdeutils-3.5.8/klaptopdaemon/portable.cpp
--- kdeutils-3.5.8/klaptopdaemon-orig/portable.cpp	2007-05-14 00:40:45.000000000 -0700
+++ kdeutils-3.5.8/klaptopdaemon/portable.cpp	2008-01-27 20:54:50.000000000 -0800
@@ -78,6 +78,7 @@
 #include <fcntl.h>
 #include <sys/ioctl.h>
 #include <dirent.h>
+#include <pwd.h>
 #include <qpushbutton.h>
 #include <qdir.h>
 #include <qfile.h>
@@ -2001,6 +2002,13 @@
 	return(0);
 }
 
+QString
+laptop_portable::get_superuser_name()
+{
+	struct passwd *entry = getpwuid(0);
+	return entry ? QString(entry->pw_name) : QString("root");
+}
+
 #elif defined(__FreeBSD__) && HAVE_MACHINE_APM_BIOS_H
 
 /*
@@ -2013,6 +2021,8 @@
 #include <stdlib.h>
 #include <machine/apm_bios.h>
 #include <sys/stat.h>
+#include <sys/types.h>
+#include <pwd.h>
 #include <qpushbutton.h>
 #include <qobject.h>
 #include <kactivelabel.h>
@@ -2426,6 +2436,12 @@
 	// software_suspend_is_preferred = hibernate;
 }
 
+QString
+laptop_portable::get_superuser_name()
+{
+	struct passwd *entry = getpwuid(0);
+	return entry ? QString(entry->pw_name) : QString("root");
+}
 
 #elif  defined(__NetBSD_APM__)
 
@@ -2435,6 +2451,8 @@
 #include <stdlib.h>
 #include <sys/ioctl.h>
 #include <sys/stat.h>
+#include <sys/types.h>
+#include <pwd.h>
 #include <machine/apmvar.h>
 #include <iostream.h>
 
@@ -2792,6 +2810,13 @@
 	values.append(s);
 }
 
+QString
+laptop_portable::get_superuser_name()
+{
+	struct passwd *entry = getpwuid(0);
+	return entry ? QString(entry->pw_name) : QString("root");
+}
+
 #else
 
 /*
@@ -3082,4 +3107,10 @@
 {
 	// software_suspend_is_preferred = hibernate;
 }
+
+static QString
+laptop_portable::get_superuser_name()
+{
+	return QString("root");
+}
 #endif
diff -Nur kdeutils-3.5.8/klaptopdaemon-orig/portable.h kdeutils-3.5.8/klaptopdaemon/portable.h
--- kdeutils-3.5.8/klaptopdaemon-orig/portable.h	2006-01-19 08:49:17.000000000 -0800
+++ kdeutils-3.5.8/klaptopdaemon/portable.h	2008-01-27 20:44:29.000000000 -0800
@@ -142,6 +142,13 @@
 	enum LaptopButton {LidButton=0, PowerButton};
 	static bool has_button(LaptopButton p);	// true if we have support for a particular button
 	static bool get_button(LaptopButton p);	// true if a button is pressed
+
+	/**
+	 * Get superuser name.
+	 * @return the superuser's name present in the passwd structure, or "root" if
+	 * an error ocurred while scanning the pw entries.
+	 */
+	static QString get_superuser_name();
 };
 
 #endif
diff -Nur kdeutils-3.5.8/klaptopdaemon-orig/sony.cpp kdeutils-3.5.8/klaptopdaemon/sony.cpp
--- kdeutils-3.5.8/klaptopdaemon-orig/sony.cpp	2006-10-01 10:28:13.000000000 -0700
+++ kdeutils-3.5.8/klaptopdaemon/sony.cpp	2008-01-27 20:41:20.000000000 -0800
@@ -118,7 +118,7 @@
 			KProcess proc;
 			proc << kdesu;
 			proc << "-u";
-			proc << "root";
+			proc << laptop_portable::get_superuser_name();
 			proc <<  "chmod +r /dev/sonypi";
 			proc.start(KProcess::Block);	// run it sync so has_acpi below sees the results
 		}
